[Donovan] iP

Donovan9617 opened the pull request:
## Duke
> I am here to make your life better.

Fast forward to 2022, Duke the chatbot is born. Duke is a 
- Text-based command-line interface
- Task tracker which helps track events and deadlines

To use Duke the chatbot, do the following:
1. Download the duke.jar file [here](https://github.com/Donovan9617/ip/releases) into an empty folder
2. Open a command window in the **same** directory as duke.jar and run the command `java -jar duke.jar` 
3. Add tasks
4. Duke the chatbot will help you manage the tasks 😊 

Here are some features of Duke the chatbot:
- [X] Add and delete tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

This is the main() function used in the chatbot:
```
    /**
     * Runs the main Duke program.
     *
     * @param args Main execution of program.
     * @throws IOException   If the filepath is incorrect.
     * @throws DukeException If the line in file contains incorrect format.
     */
    public static void main(String[] args) throws IOException, DukeException {
        Duke duke = new Duke("data/tasks.txt");
        duke.run();
    }
```

Donovan9617 committed:
Level-1: Greet, Echo, Exit

Donovan9617 committed:
Level-2: Add, List

Donovan9617 committed:
Level-3: Mark as Done

Donovan9617 committed:
Level-4: ToDo, Event, Deadline

Donovan9617 committed:
A-TextUiTesting: Automated Text UI Testing

Donovan9617 committed:
Level-5: Handle Errors

Donovan9617 committed:
Level-6: Delete

Donovan9617 committed:
A-Enums: Use Enums

Donovan9617 committed:
Update to coding standards for Packages and Switch statement

Donovan9617 committed:
Level 7

Donovan9617 committed:
Level 8: Dates and Times

Donovan9617 committed:
A-MoreOOP

Donovan9617 committed:
Reformatted date

Donovan9617 committed:
A-Packages: Organize into Packages

Donovan9617 committed:
A-Gradle: Use Gradle

Donovan9617 committed:
A-JUnit: Add JUnit Tests

Donovan9617 committed:
A-Jar: Create a JAR File

Donovan9617 committed:
A-JavaDoc

Donovan9617 committed:
Merging

Donovan9617 committed:
Merge branch 'branch-A-JavaDoc'

Donovan9617 committed:
Coding Standards

Donovan9617 committed:
Coding standards formatting

Donovan9617 committed:
Level-9: Find

Donovan9617 committed:
Editted file formats to create JAR file

Donovan9617 committed:
A-Checkstyle: Use Checkstyle

shaniceng commented on src/main/java/Duke/Task.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @return String tick if task is done else an empty space.
+     */
+    public String getStatusIcon() {
```
Great job in naming all the methods in verb and in camel cases 👍

shaniceng commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,91 @@
+ */
+public class TaskList {
+
+    private ArrayList<Task> listOfTasks;
```
Suggest to name this as tasks instead, just to keep things short and simple
```suggestion
    private ArrayList<Task> tasks;

```

Donovan9617 replied:
Thanks!

zupey commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+
+    private Ui ui;
+    private Storage storage;
+    private TaskList tasks;
```
Should this be better named as a singular because TaskList is technically not a collection of tasks like ArrayList<Task>?

Donovan9617 replied:
Thanks for highlighting that

Donovan9617 replied:
I see, thanks for the suggestion!

zupey commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+    public void run() throws IOException {
+        ui.showWelcome();
+        boolean stopRunning = false;
+        while (!stopRunning) {
```
Did you intend for this loop to never exit because I can't seem to find where this variable might be changed?

Donovan9617 replied:
Thanks for highlighting that!

zupey commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,135 @@
+            throws DukeException, IOException {
+        String[] response = command.split(" ");
+        Response firstWord = Response.valueOf(response[0].toUpperCase());
+        switch (firstWord) {
```
Great that you followed that there should be no indentation for `case` clauses. 👍

zupey commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,142 @@
+     * @param file The specified path to the file.
+     * @throws DukeException when a subdirectory or the file is unable to be created.
+     */
+    private static void createFileAndSubdirectoriesIfFileNotFound(File file) throws DukeException {
```
Method name is very descriptive of its tasks! 👍

zupey reviewed (commented):
LGTM

shaniceng commented on src/main/java/Duke/Ui.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    /**
+     * Prints welcome message to console.
```
Well done in giving all the methods comments and it is clear what these methods are for 👍

shaniceng commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,23 @@
+public class TaskListTest {
+
+    @Test
+    public void add_task_taskAdded() {
```
Even though this is a test case but should these methods be named as camel cases instead? I noticed the same issue with methods naming in other test cases as well.
```suggestion
    public void addTaskTaskAdded() {

```

Donovan9617 replied:
Thanks for highlighting!

shaniceng commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * Write taskList to duke.txt.
```
Should this be "Writes" instead of "Write" by following the convention of java coding standard?
```suggestion
     * Writes taskList to duke.txt.

```

Donovan9617 replied:
Thanks!

shaniceng reviewed (commented):
Overall, your code looks good and it LGTM 👍 
Just some nits to be fixed. Good Job!

Donovan9617 committed:
Level-10: GUI

Donovan9617 committed:
Edits to test and tasklist variable namings

Donovan9617 committed:
Added java docs for test classes

crvstalphua commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,157 @@
+    public Duke() {
+        ui = new Ui();
+        storage = new Storage("/data/tasks.txt");
+        TaskList currentTasks;
```
Similar to the another comment, perhaps this could also be singular since TaskList is technically one object which is a list of tasks? 
```suggestion
        TaskList currentTaskList;

```

Donovan9617 replied:
Thanks for highlighting!

crvstalphua commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,91 @@
+ */
+public class TaskList {
+
+    private ArrayList<Task> tasks;
```
Great job using plural name for ArrayList which is a collection of tasks!

crvstalphua commented on src/main/java/Duke/Ui.java:
```diff
@@ -0,0 +1,205 @@
+package duke;
```
Despite having many different methods in this class, it was still very clear and easy to read because method names were all clear (in camelCase) with javadoc headers for explanations!

crvstalphua commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,143 @@
+            }
+            break;
+
+        case OTHERS:
```
If I recall correctly, it is recommended to just use default for the final enumeration case, not that it would really make a difference but perhaps for consistent coding style? :")

crvstalphua commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,32 @@
+     * Tests for adding a Task object to TaskList.
+     */
+    @Test
+    public void addTaskTaskAdded() {
```
For your test cases (in other documents as well), perhaps you could follow this naming format indicated in the coding standards:

featureUnderTest_testScenario_expectedBehavior()
```suggestion
    public void addTask_TaskAdded() {

```

Donovan9617 replied:
Thanks for the suggestion!

crvstalphua reviewed (commented):
Overall, most of the methods had comments and followed the naming standards as far as I could tell. LGTM, other than some nits that I've suggested! :D

Donovan9617 committed:
Changes to naming of test methods

Donovan9617 committed:
Parser class: Add assert feature to methods to document important
aspects that should hold at various points of the code
TaskList class: Add assert feature to methods to document important
aspects that should hold at various points of the code

Documenting aspects that should hold at various points ensures the
program is running correctly at intermediate stages.

When assert runs, it checks if the condition provided (some desired
state of the code) is satisfied, else it will output the given string
text.

Using assertions helps to ensure program runs correctly at intermediate
stages, and detect incorrect intermediates and outputs quickly.

Refer to the link here to learn more about
assertions: https://www.geeksforgeeks.org/assertions-in-java/

Donovan9617 committed:
Merge pull request #1 from Donovan9617/branch-A-Assertions

Assert features

Donovan9617 committed:
Parser class: use string type for switch variable over enum type.

Needs to change as a substring that is not found in enum will cause an error.

Use string type for switch variable instead of the enum type. Switch
cases are in string too.

This enables the bot to catch unrecognised commands and contain them in
its reply more effectively.

Donovan9617 committed:
Merge pull request #2 from Donovan9617/branch-A-CodeQuality

Improved Code Quality for Parser and other classes

Donovan9617 committed:
Setting up custom workflow

Donovan9617 committed:
Added CI by setting up .github folder

Donovan9617 committed:
Added CI

Donovan9617 committed:
Changing permission of gradlew

Donovan9617 committed:
Removed trailing whitespace

Donovan9617 committed:
Added BetterSearch to search with partial keywords

Donovan9617 committed:
added C-Undo feature to Duke

Donovan9617 committed:
Added java docs for method in TaskList class

Donovan9617 committed:
Merge pull request #3 from Donovan9617/branch-C-Undo

Branch c undo

Donovan9617 committed:
Change code format in Storage and Duke

Donovan9617 committed:
Change line formatting

Donovan9617 committed:
Remove redundant code

Donovan9617 committed:
Add screenshot of product to docs folder

Donovan9617 committed:
Update README document

Donovan9617 committed:
Merge pull request #5 from Donovan9617/A-UserGuide

Update README document

Donovan9617 committed:
Update user guide

Donovan9617 committed:
Merge pull request #6 from Donovan9617/branch-A-UserGuide

Update user guide

Donovan9617 committed:
Add console display for chat bot

Donovan9617 committed:
Format code

Donovan9617 committed:
Add display for no tasks

Donovan9617 committed:
Add error handlings

Donovan9617 committed:
Add design to GUI

Donovan9617 committed:
SLAP for Parser class

Donovan9617 committed:
Edit Ui image
